src/ui/mod.rs

```diff
@@ -90,48 +86,50 @@ impl TakeTurnState {
+                                                                 &font,
+                                                                 |idx, card, name| Some(GameEvent::CardBought(name, idx)));
+            widgets.push(Box::new(shop_zone));
         }
 
         // buildings
         let mut base_playzone_position = Vector::new(60.0 * UI_UNIT, PLAYER_BOARD_FROM_TOP);
 
-        let mut zone: CardZone<CardIcon> = CardZone::new(
-            String::from("Buildings"),
-            base_playzone_position,
-            ZoneDirection::Vertical,
-        );
-        for (num, card) in gameplay_state.get_board().buildings.list.clone().drain(..).enumerate() {
-            zone.add(card, &font, None)
-        }
-        widgets.push(Box::new(zone));
+        let build_zone = CardZone::<CardIcon>::from_container(&gameplay_state.get_board().buildings,
+                                                              base_playzone_position,
+                                                              ZoneDirection::Vertical,
+                                                              &font,
+                                                              |idx, card, name| None);
+        widgets.push(Box::new(build_zone));
 
         // kaiju_zone
+
         let mut zone: CardZone<CardIcon> = CardZone::new(
-            String::from("Kaiju zone"),
+            BoardZone::Kaiju,
             base_playzone_position + Vector::new(UI_UNIT * 9.0, 0), // 7U widget height + 1U padding + 1U gap
             ZoneDirection::Vertical,
         );
 
-        for (num, card) in gameplay_state.get_board().kaiju_zone.clone().drain(..).enumerate() {
+        for (num, card) in gameplay_state.get_board().kaiju_zone.cards.clone().drain(..).enumerate() {
             zone.add(card, &font, None)
         }
+
+        let kaiju_position = base_playzone_position + Vector::new(UI_UNIT * 9.0, 0); // 7U widget height + 1U padding + 1U gap
+        let kaiju_zone = CardZone::<CardIcon>::from_container(&gameplay_state.get_board().kaiju_zone,
+                                                              kaiju_position,
+                                                              ZoneDirection::Vertical,
+                                                              &font,
+                                                              |idx, card, zone_id| None);
         widgets.push(Box::new(zone));
 
         let base_numbers_position = Vector::new(4.0 * UI_UNIT, PLAYER_BOARD_FROM_TOP + 15.0 * UI_UNIT);
-        for (num, (name, value)) in gameplay_state.get_board().globals.iter().enumerate() {
+
+        for (num, currency) in Globals::in_game().iter().enumerate() {
+            let value = gameplay_state.get_board().globals.get(*currency);
             widgets.push(Box::new(Button::new(
-                format!("{}\n {}", name, value),
+                format!("{:?}\n {}", currency, value),
```

Comment:
This uses Debug formatting for user-facing interface. This is wrong: you should implement Display. 
It is probably OK to implement it just as 

```
fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
    // We know that debug formatting is OK
    <self as Debug>::fmt(f)
}
```

Reply:
BTW: here you can see the rare left-swimming turbofish!

Reply:
ok, I'll use this